File: indexer/packages/redis/src/scripts/place_order.lua
-- TODO(CORE-512): add info/resources around caches.
-- Doc: https://www.notion.so/dydx/Indexer-Technical-Spec-a6b15644502048f994c98dee35b96e96#61d5f8ca5117476caab78b3f0691b1d0

-- Key for the value in the orders cache, see `src/caches/orders-cache.ts` for more details
local orderKey = KEYS[1];
-- Key for the value in the orders data cache, see `src/caches/orders-data-cache.ts` for more
-- details
local orderDataKey = KEYS[2];
-- Key for the value in the subaccount order ids cache, see
-- `src/caches/subaccount-order-ids-cache.ts` for more details
local subaccountKey = KEYS[3];
-- Key for the orderExpiry cache, see `src/caches/order-expiry-cache.ts` for more details
local expiryKey = KEYS[4];

-- Order that is being placed, in encoded proto form
local newOrder = ARGV[1];
-- Expiry (good-til-block or sequence number) of the order being placed
local newOrderExpiry = ARGV[2];
-- UUID of the order being placed
local orderId = ARGV[3];
-- Whether the order that is being placed is a short-term order. Need to convert from string to bool
local isShortTermOrder = ARGV[4] == "true";

-- This script returns the following values in an array
-- 1. Was a new order placed or replaced - 1 if an order is placed, 0 if not
-- 2. Was an order replaced - 1 if an order was replaced, 0 if not
-- 3. Total filled size of the old order in quantums - "0" if an order was not replaced
-- 4. Was the old order resting on the book - "true/false", "false" if an order was not replaced
-- 5. Encoded old order - "" if an order was not replaced

-- order data has the format:
-- [good-til-block or sequence number of order]_[total filled]_[true/false, if order is on the book]

-- check if the order exists
local oldOrder = redis.call("get", orderKey);
if not oldOrder then
  -- create the order, set up the order data with total filled size = 0, "false" for whether the
  -- order is resting on the book and add order to list of orders for the subaccount
  redis.call("set", orderKey, newOrder);
  -- refer to above comment on order data format
  redis.call("set", orderDataKey, newOrderExpiry .. "_0_false");
  redis.call("hset", subaccountKey, orderId, 1);
  -- Long-term orders will be on-chain, so we only need to store expiry data for short-term orders
  if isShortTermOrder then
    redis.call("zadd", expiryKey, newOrderExpiry, orderId);
  end
  return {1, 0, "0", "false", ""};
else
  -- refer to above comment on order data format
  local oldOrderData = redis.call("get", orderDataKey);
  local j = string.find(oldOrderData, "_");
  local oldExpiry = string.sub(oldOrderData, 1, j - 1);
  local i = string.find(oldOrderData, "_", j + 1)
  local oldTotalFilledQuantums = string.sub(oldOrderData, j + 1, i - 1);
  local oldRestingOnBook = string.sub(oldOrderData, i + 1);

  -- if the new order has a lower or equal expiry (good-til-block or sequence number) than the order
  -- in the cache, return early
  if tonumber(oldExpiry) >= tonumber(newOrderExpiry) then
    return {0, 0, "0", "false", ""}
  end

  -- update the order if the new order has a greater expiry (good-til-block or sequence number) than
  -- the order in the cache, also update the order data with the new expiry and the total filled
  -- of the older order. As the order is replaced, it is no longer resting on the book, so set the
  -- to "false".
  redis.call("set", orderKey, newOrder)
  -- refer to the above comment on order data format
  redis.call("set", orderDataKey, newOrderExpiry .. "_" .. oldTotalFilledQuantums .. "_false")
  -- Long-term orders will be on-chain, so we only need to store expiry data for short-term orders
  if isShortTermOrder then
    -- The expiry is guaranteed to be different, so overwrite the old one from the expiry cache
    redis.call("zadd", expiryKey, newOrderExpiry, orderId)
  end

  -- if the new order replaced an older order, return data about the old order to use to update
  -- the orderbook
  return {0, 1, oldTotalFilledQuantums, oldRestingOnBook, oldOrder}
end

File: indexer/packages/redis/src/caches/place-order.ts
import { OrderTable } from '@dydxprotocol-indexer/postgres';
import { ORDER_FLAG_SHORT_TERM } from '@dydxprotocol-indexer/v4-proto-parser';
import { RedisOrder } from '@dydxprotocol-indexer/v4-protos';
import { Callback, RedisClient } from 'redis';

import { PlaceOrderResult } from '../types';
import { getOrderExpiry, validateRedisOrder } from './helpers';
import { ORDER_EXPIRY_CACHE_KEY } from './order-expiry-cache';
import { getOrderCacheKey } from './orders-cache';
import { getOrderDataCacheKey } from './orders-data-cache';
import { placeOrderScript } from './scripts';
import { getSubaccountOrderIdsCacheKey } from './subaccount-order-ids-cache';

// Number of keys that the lua script will access.
const numKeys: number = 4;

/**
 * Updates order caches in Redis for an order being placed/replaced. Evaluates the `place_order.lua`
 * script to update the caches.
 * The behavior of this should be:
 * - if the order is new:
 *   - the encoded `RedisOrder` should be saved to the `ORDERS_CACHE`, mapped to the order uuid
 *   - {good-til-block/sequence-number}_{totalFilled}_false should be saved to the
 *     `ORDERS_DATA_CACHE,` mapped to the order uuid, `totalFilled` being set to 0
 *   - the order uuid should be added to the `SUBACCOUNT_ORDERS_CACHE` for the subaccount that
 *     placed the order
 *   - if the order is a short-term order:
 *     - the order uuid should be added to the `ORDER_EXPIRY_CACHE` with its score set to its expiry
 * - if the order exists in the cache, and the placed order has a lower or equal expiry
 *   (good-til-block/seq. number)
 *   - no updates
 * - if the order exists in the cache, and the placed order has a greater expiry
 *   - the encoded `RedisOrder` should be saved to the `ORDERS_CACHE`, replacing the existing order
 *   - {good-til-block/sequence-number}_{totalFilled}_false should be saved to the
 *     `ORDERS_DATA_CACHE`, replacing the existing order, with `totalFilled` being the value for the
 *     existing order
 *   - if the order is a short-term order:
 *     - the score for the order uuid in the `ORDER_EXPIRY_CACHE` should be updated to the new
 *       expiry
 * See the `place_order.lua` script for more context.
 * @param param0 Contains the `RedisOrder` of the order being placed.
 * @returns `PlaceOrderResult` for the result of placing the order.
 */
export async function placeOrder({
  redisOrder,
  client,
}: {
  redisOrder: RedisOrder,
  client: RedisClient,
}): Promise<PlaceOrderResult> {
  validateRedisOrder(redisOrder);

  let evalAsync: (
    orderKey: string,
    orderDataKey: string,
    subaccountOrderIdsKey: string,
    orderExpiryKey: string,
    encodedOrder: string,
    orderExpiry: string,
    orderId: string,
    isShortTermOrder: boolean,
  ) => Promise<PlaceOrderResult> = (
    orderKey,
    orderDataKey,
    subaccountOrderIdsKey,
    orderExpiryKey,
    encodedOrder,
    orderExpiry,
    orderId,
    isShortTermOrder,
  ) => {
    return new Promise((resolve, reject) => {
      const callback: Callback<[number, number, string, string, string]> = (
        err: Error | null,
        results: [number, number, string, string, string],
      ) => {
        if (err) {
          return reject(err);
        }
        const [
          placed,
          replaced,
          filledStr,
          bookStr,
          oldOrder,
        ]:[
          number,
          number,
          string,
          string,
          string,
        ] = results;
        return resolve({
          placed: placed === 1,
          replaced: replaced === 1,
          oldTotalFilledQuantums: replaced === 1 ? parseInt(filledStr, 10) : undefined,
          restingOnBook: replaced === 1 ? bookStr === 'true' : undefined,
          oldOrder: replaced === 1 ? RedisOrder.decode(Buffer.from(oldOrder, 'binary')) : undefined,
        });
      };
      client.evalsha(
        placeOrderScript.hash,
        numKeys,
        orderKey,
        orderDataKey,
        subaccountOrderIdsKey,
        orderExpiryKey,
        encodedOrder,
        orderExpiry,
        orderId,
        isShortTermOrder.toString(),
        callback,
      );
    });
  };
  evalAsync = evalAsync.bind(client);

  return evalAsync(
    getOrderCacheKey(redisOrder.order!.orderId!),
    getOrderDataCacheKey(redisOrder.order!.orderId!),
    getSubaccountOrderIdsCacheKey(redisOrder.order!.orderId!.subaccountId!),
    ORDER_EXPIRY_CACHE_KEY,
    // TODO: use String to directly convert the UInt8Array to a string
    Buffer.from(RedisOrder.encode(redisOrder).finish()).toString('binary'),
    getOrderExpiry(redisOrder.order!).toString(),
    OrderTable.orderIdToUuid(redisOrder.order!.orderId!),
    redisOrder.order!.orderId!.orderFlags === ORDER_FLAG_SHORT_TERM,
  );
}
